src/supervised/supervised.jl

```diff
@@ -36,141 +36,150 @@ islipschitzcont_deriv(::SupervisedLoss) = false
-    n = size(output, 2)
-    k = size(output, 1)
-    @_dimcheck size(target) == size(output) && size(buffer) == (k, n)
-    for i = 1:n
-        @simd for j = 1:k
-            @inbounds buffer[j, i] = deriv(loss, target[j, i], output[j, i])
-        end
-    end
-    buffer
-end
-
-# --------------------------------------------------------------
-
-function sumvalue{T<:Number}(loss::SupervisedLoss, target::AbstractVector, output::AbstractArray{T})
-    n = length(output)
-    @_dimcheck length(target) == n
-    val = zero(T)
-    @simd for i = 1:n
-        @inbounds val += value(loss, target[i], output[i])
+@generated function sumvalue{T,N,Q,M}(
+        loss::SupervisedLoss,
+        target::AbstractArray{Q,M},
+        output::AbstractArray{T,N}
+    )
+    quote
+      @_dimcheck size(buffer) == size(output)
+      @nexprs $N (n)->(s_n = size(output,n))
+      @nexprs $M (n)->@_dimcheck(size(target,n) == s_n)
+      val = zero(T) # TODO: this might be not be type-stable?
+      @simd for I in CartesianRange(@ntuple($N,n->(1:s_n)))
+          @nexprs $N n->(i_n = I[n])
+          @inbounds val += value(loss, @nref($M,target,i), @nref($N,output,i))
+      end
+      val
     end
-    val
 end
 
-function sumderiv{T<:Number}(loss::SupervisedLoss, target::AbstractVector, output::AbstractArray{T})
-    n = length(output)
-    @_dimcheck length(target) == n
-    val = zero(T)
-    @simd for i = 1:n
-        @inbounds val += deriv(loss, target[i], output[i])
+@generated function sumderiv{T,N,Q,M}(
+        loss::SupervisedLoss,
+        target::AbstractArray{Q,M},
+        output::AbstractArray{T,N}
+    )
+    quote
+      @_dimcheck size(buffer) == size(output)
+      @nexprs $N (n)->(s_n = size(output,n))
+      @nexprs $M (n)->@_dimcheck(size(target,n) == s_n)
+      val = zero(T) # TODO: this might be not be type-stable?
+      @simd for I in CartesianRange(@ntuple($N,n->(1:s_n)))
+          @nexprs $N n->(i_n = I[n])
+          @inbounds val += deriv(loss, @nref($M,target,i), @nref($N,output,i))
+      end
+      val
     end
-    val
 end
 
 # --------------------------------------------------------------
 
-function meanvalue{T<:Number}(loss::SupervisedLoss, target::AbstractVector, output::AbstractArray{T})
-    n = length(output)
-    @_dimcheck length(target) == n
-    val = zero(T)
-    tmp = zero(T)
-    @simd for i = 1:n
-        @inbounds tmp = value(loss, target[i], output[i])::T
-        tmp /= n
-        val += tmp
+@generated function meanvalue{T,N,Q,M}(
+        loss::SupervisedLoss,
+        target::AbstractArray{Q,M},
+        output::AbstractArray{T,N}
+    )
+    quote
+      @_dimcheck size(buffer) == size(output)
+      @nexprs $N (n)->(s_n = size(output,n))
+      @nexprs $M (n)->@_dimcheck(size(target,n) == s_n)
+      val = zero(T) # TODO: this might be not be type-stable?
+      tmp = zero(T) # TODO: this might be not be type-stable?
```

Comment:
I'm ok with this for now. Just leave the todo there. I would like to keep this code simple for now until I got some time to play with it